### PR TITLE
Pal 963 Kafka health indicator for topic offset service

### PR DESCRIPTION
**Pull request by m78233**

### CODE REVIEW

File: topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/kafka/KafkaContractTest.java [hunk 1]
```diff
@@ -85,36 +85,37 @@
 
 /**
  * An external requirement of the service is to connect to a pair of kafka topics.
- * The upstream "masked-resource" topic is written to by the attribute-masking-service and read by this service.
- * The downstream "masked-resource-offset" topic is written to by this service and read by the filtered-resource-service.
+ * The upstream "attribute-masking-service" topic is and read by this service.
+ * The downstream "filtered-resource-offset" topic is written to by this service.
```
Comment (developer01189998819991197253) [inline topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/kafka/KafkaContractTest.java]: The names included in the old diff are the names of the topics, not the names of the services, revert those changes

File: topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/rest/HealthActuatorContractTest.java [hunk 2]
```diff
@@ -30,31 +30,31 @@
 import static org.assertj.core.api.Assertions.assertThat;
 
 /**
- * An external requirement of the service is to keep-alive in k8s.
- * This is done by checking the service is still alive and healthy by REST GET /actuator/health.
- * This should return 200 OK if the service is healthy.
+ * SpringBoot Starter Actuator is a service loaded-up by adding the spring-boot-starter-actuator as a dependency to the
+ * project and configured in the application.yaml file.  It is a service which provides information on the application
+ * and is being used to monitor the "health" of the palisade service.  If there is an indication that this service has
+ * fallen over, this information can be used to restore the service.
```
Comment (developer01189998819991197253) [inline topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/rest/HealthActuatorContractTest.java]: We dont need to explain what the SpringBoot starter actuator is, saying that this test checks the services health endpoint is enough

File: topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/rest/HealthActuatorContractTest.java [hunk 3]
```diff
@@ -30,31 +30,31 @@
...
-        // When we GET the /actuator/health REST endpoint (used by k8s)
-        final ResponseEntity<String> health = restTemplate.getForEntity("/actuator/health", String.class);
-
-        // Then the service reports itself to be healthy
-        assertThat(health.getStatusCodeValue()).isEqualTo(HttpStatus.OK.value());
+    void testHealthActuatorServiceIsRunning() {
+        ResponseEntity<String> responseEntity = restTemplate.getForEntity("/actuator/health", String.class);
+        assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
+        String body = responseEntity.getBody();
+        assertThat(body).contains("\"status\":\"UP\"");
```
Comment (developer01189998819991197253) [inline topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/rest/HealthActuatorContractTest.java]:
```suggestion
        assertAll("Assert the Health Actuator",
                () -> assertThat(responseEntity.getStatusCode())
                        .as("Check the status code of the response")
                        .isEqualTo(HttpStatus.OK),

                () -> assertThat(responseEntity.getBody())
                        .as("Check the body of the response")
                        .contains("\"status\":\"UP\"")
        );
```

Comment (developer01189998819991197253) [reply topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/rest/HealthActuatorContractTest.java]: grouped assertion works a bit better here

File: topic-offset-service/src/main/java/uk/gov/gchq/palisade/service/topicoffset/service/KafkaHealthIndicator.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                return consumerGroupDescription.members().stream()
+                        .noneMatch(member -> (member.assignment() == null || member.assignment().topicPartitions().isEmpty()));
+            }
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            LOGGER.warn("Timeout during Kafka health check for group {}", this.groupId, e);
+            Thread.currentThread().interrupt();
+            return false;
+        }
+        return true;
+    }
```
Comment (developer01189998819991197253) [inline topic-offset-service/src/main/java/uk/gov/gchq/palisade/service/topicoffset/service/KafkaHealthIndicator.java]:
```suggestion
    private boolean performCheck() {
        try {
            Map<String, ConsumerGroupDescription> groupDescriptionMap = this.adminClient.describeConsumerGroups(Collections.singletonList(this.groupId))
                    .all()
                    .get(1, TimeUnit.SECONDS);

            ConsumerGroupDescription consumerGroupDescription = groupDescriptionMap.get(this.groupId);
            LOGGER.debug("Kafka consumer group ({}) state: {}", groupId, consumerGroupDescription.state());

            if (consumerGroupDescription.state() == ConsumerGroupState.STABLE) {
                boolean assignedGroupPartition = consumerGroupDescription.members().stream()
                        .noneMatch(member -> (member.assignment() == null || member.assignment().topicPartitions().isEmpty()));
                if (!assignedGroupPartition) {
                    LOGGER.error("Failed to find kafka topic-partition assignments");
                }
                return assignedGroupPartition;
            }

        } catch (InterruptedException e) {
            LOGGER.warn("Await on future interrupted", e);
            Thread.currentThread().interrupt();
        } catch (ExecutionException | TimeoutException e) {
            LOGGER.warn("Timeout connecting to kafka", e);
        }
        return false;
    }
```

Comment (developer01189998819991197253) [reply topic-offset-service/src/main/java/uk/gov/gchq/palisade/service/topicoffset/service/KafkaHealthIndicator.java]: Other than some whitespace changes, the main diff here is the split out catch block at the bottom that allows for seperate error handling

File: topic-offset-service/src/main/java/uk/gov/gchq/palisade/service/topicoffset/stream/ProducerTopicConfiguration.java [hunk 5]
```diff
@@ -64,8 +64,7 @@ public Integer getParallelism() {
     @Generated
     public void setParallelism(final Integer parallelism) {
         this.parallelism = Optional.ofNullable(parallelism)
-                .filter(x -> x > 0)
-                .orElseThrow(() -> new IllegalArgumentException("parallelism must be positive non-null"));
+                .orElseThrow(() -> new IllegalArgumentException("parallelism cannot be null"));
```
Comment (developer01189998819991197253) [inline topic-offset-service/src/main/java/uk/gov/gchq/palisade/service/topicoffset/stream/ProducerTopicConfiguration.java]: we want the filter in here
```
this.parallelism = Optional.ofNullable(parallelism)
                .filter(x -> x > 0)
                .orElseThrow(() -> new IllegalArgumentException("parallelism must be positive non-null"));

**Review by developer01189998819991197253 — CHANGES_REQUESTED** [review]
Few small changes

Also missing implementation within the remaining services

File: topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/kafka/KafkaContractTest.java (hunk 1 above)
Comment (m78233) [reply topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/kafka/KafkaContractTest.java]: Ok

File: topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/rest/HealthActuatorContractTest.java (hunk 2 above)
Comment (m78233) [reply topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/rest/HealthActuatorContractTest.java]: Ok

File: topic-offset-service/src/main/java/uk/gov/gchq/palisade/service/topicoffset/stream/ProducerTopicConfiguration.java (hunk 5 above)
Comment (m78233) [reply topic-offset-service/src/main/java/uk/gov/gchq/palisade/service/topicoffset/stream/ProducerTopicConfiguration.java]: Ok

File: topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/kafka/KafkaContractTest.java [hunk 6]
```diff
@@ -85,8 +85,8 @@
 
 /**
  * An external requirement of the service is to connect to a pair of kafka topics.
- * The upstream "attribute-masking-service" topic is and read by this service.
- * The downstream "filtered-resource-offset" topic is written to by this service.
+ * The upstream "masked-resource" topic is written to by the attribute-masking-service and read by this service.
+ * The downstream "masked-resource-offset" topic is written to by this service and read by the filtered-resource-service.
```
Comment (developer01189998819991197253) [inline topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/kafka/KafkaContractTest.java]:
```suggestion
 * The downstream "masked-resource-offset" topic is written to by this service and read by the Filtered-Resource-Service.
```

File: topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/rest/HealthActuatorContractTest.java [hunk 7]
```diff
@@ -28,12 +28,10 @@
...
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertAll;
 
 /**
- * SpringBoot Starter Actuator is a service loaded-up by adding the spring-boot-starter-actuator as a dependency to the
- * project and configured in the application.yaml file.  It is a service which provides information on the application
- * and is being used to monitor the "health" of the palisade service.  If there is an indication that this service has
- * fallen over, this information can be used to restore the service.
+ * Test checks the services health endpoint for the topic-offset-service
```
Comment (developer01189998819991197253) [inline topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/rest/HealthActuatorContractTest.java]:
```suggestion
 * Test checks the services health endpoint for the Topic-Offset-Service
```

File: topic-offset-service/src/main/java/uk/gov/gchq/palisade/service/topicoffset/stream/ProducerTopicConfiguration.java [hunk 8]
```diff
@@ -64,6 +64,7 @@ public Integer getParallelism() {
     @Generated
     public void setParallelism(final Integer parallelism) {
         this.parallelism = Optional.ofNullable(parallelism)
+                .filter(x -> x > 0)
                 .orElseThrow(() -> new IllegalArgumentException("parallelism cannot be null"));
```
Comment (developer01189998819991197253) [inline topic-offset-service/src/main/java/uk/gov/gchq/palisade/service/topicoffset/stream/ProducerTopicConfiguration.java]:
```suggestion
               .orElseThrow(() -> new IllegalArgumentException("parallelism must be positive non-null"));
```

**Review by developer01189998819991197253 — CHANGES_REQUESTED** [review]
bit of javadoc

File: topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/kafka/KafkaContractTest.java (hunk 6 above)
Comment (m78233) [reply topic-offset-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/topicoffset/kafka/KafkaContractTest.java]: ok

**Review by developer01189998819991197253 — DISMISSED** [review]

**Review by kg981167224 — DISMISSED** [review]

File: attribute-masking-service/src/main/java/uk/gov/gchq/palisade/service/attributemask/service/KafkaHealthIndicator.java [hunk 9]
```diff
@@ -38,29 +37,33 @@
...
-        return Optional.of(performCheck())
-                .filter(healthy -> healthy)
-                .map(up -> Health.up().withDetail("group", this.groupId).build())
-                .orElseGet(() -> Health.down().withDetail("group", this.groupId).build());
+        if (performCheck()) {
+            return Health.up().withDetail("group", this.groupId).build();
+        } else {
+            return Health.down().withDetail("group", this.groupId).build();
+        }
     }
```
Comment (dev930018) [inline attribute-masking-service/src/main/java/uk/gov/gchq/palisade/service/attributemask/service/KafkaHealthIndicator.java]: Don't override this, spring will provide a better default (i.e. actually using the includeDetails flag).
See `org.springframework.boot.actuate.health.HealthIndicator` class:
```
public interface HealthIndicator extends HealthContributor {

	/**
	 * Return an indication of health.
	 * @param includeDetails if details should be included or removed
	 * @return the health
	 * @since 2.2.0
	 */
	default Health getHealth(boolean includeDetails) {
		Health health = health();
		return includeDetails ? health : health.withoutDetails();
	}

	/**
	 * Return an indication of health.
	 * @return the health
	 */
	Health health();

}
```
You should only implement the health() method

File: attribute-masking-service/src/main/java/uk/gov/gchq/palisade/service/attributemask/service/KafkaHealthIndicator.java [hunk 10]
```diff
@@ -70,10 +73,11 @@ public Health getHealth(final boolean includeDetails) {
...
     public Health health() {
-        return Optional.of(performCheck())
-                .filter(healthy -> healthy)
-                .map(up -> Health.up().build())
-                .orElseGet(() -> Health.down().build());
+        if (performCheck()) {
+            return Health.up().build();
+        } else {
+            return Health.down().build();
+        }
```
Comment (dev930018) [inline attribute-masking-service/src/main/java/uk/gov/gchq/palisade/service/attributemask/service/KafkaHealthIndicator.java]: Include the groupId detail? You should put as much detail as necessary in here, then spring will use its default `getHealth(bool includeDetails)` method, which will remove your details added if they are not required by config/request-type/etc.

File: topic-offset-service/src/main/java/uk/gov/gchq/palisade/service/topicoffset/stream/ProducerTopicConfiguration.java [hunk 11]
```diff
@@ -65,7 +65,7 @@ public Integer getParallelism() {
     public void setParallelism(final Integer parallelism) {
         this.parallelism = Optional.ofNullable(parallelism)
                 .filter(x -> x > 0)
-                .orElseThrow(() -> new IllegalArgumentException("parallelism must be positive non-null"));
+               .orElseThrow(() -> new IllegalArgumentException("parallelism must be positive non-null"));
```
Comment (dev930018) [inline topic-offset-service/src/main/java/uk/gov/gchq/palisade/service/topicoffset/stream/ProducerTopicConfiguration.java]: Revert this change
```suggestion
                .orElseThrow(() -> new IllegalArgumentException("parallelism must be positive non-null"));
```

**Review by dev930018 — CHANGES_REQUESTED** [review]
The couple of changes to the kafka health indicator should apply to all services.

File: attribute-masking-service/src/main/java/uk/gov/gchq/palisade/service/attributemask/service/KafkaHealthIndicator.java (hunk 9 above)
Comment (m78233) [reply attribute-masking-service/src/main/java/uk/gov/gchq/palisade/service/attributemask/service/KafkaHealthIndicator.java]: Removed the override method Health getHealth(final boolean includeDetails) from each of the services.

File: attribute-masking-service/src/main/java/uk/gov/gchq/palisade/service/attributemask/service/KafkaHealthIndicator.java (hunk 10 above)
Comment (m78233) [reply attribute-masking-service/src/main/java/uk/gov/gchq/palisade/service/attributemask/service/KafkaHealthIndicator.java]: changed the code for each of the services

@Override
    public Health health() {
        if (performCheck()) {
            return Health.up().withDetail("group", this.groupId).build();
        } else {
            return Health.down().withDetail("group", this.groupId).build();
        }
    }

**Review by dev930018 — APPROVED** [review]

**Review by developer01189998819991197253 — APPROVED** [review]
